File: src/problems/add_two_numbers.rs
use Box;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub struct Solution;

impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut l1 = l1;
        let mut l2 = l2;
        let mut carry = 0;
        let mut dummy = Some(Box::new(ListNode::new(0)));
        let mut current = dummy.as_mut();

        while l1.is_some() || l2.is_some() {
            let mut sum = carry;

            if let Some(node) = l1 {
                sum += node.val;
                l1 = node.next;
            }

            if let Some(node) = l2 {
                sum += node.val;
                l2 = node.next;
            }

            carry = sum / 10;
            current.as_mut().unwrap().next = Some(Box::new(ListNode::new(sum % 10)));
            current = current.and_then(|node| node.next.as_mut());
        }

        if carry > 0 {
            current.as_mut().unwrap().next = Some(Box::new(ListNode::new(carry)));
        }

        dummy.unwrap().next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        let l1 = Some(Box::new(ListNode {
            val: 2,
            next: Some(Box::new(ListNode {
                val: 4,
                next: Some(Box::new(ListNode::new(3))),
            })),
        }));
        let l2 = Some(Box::new(ListNode {
            val: 5,
            next: Some(Box::new(ListNode {
                val: 6,
                next: Some(Box::new(ListNode::new(4))),
            })),
        }));
        let result = Solution::add_two_numbers(l1, l2);
        assert_eq!(
            result,
            Some(Box::new(ListNode {
                val: 7,
                next: Some(Box::new(ListNode {
                    val: 0,
                    next: Some(Box::new(ListNode::new(8))),
                })),
            }))
        );
    }

    #[test]
    fn test_2() {
        let l1 = Some(Box::new(ListNode::new(0)));
        let l2 = Some(Box::new(ListNode::new(0)));
        let result = Solution::add_two_numbers(l1, l2);
        assert_eq!(result, Some(Box::new(ListNode::new(0))));
    }

    #[test]
    fn test_3() {
        let l1 = Some(Box::new(ListNode {
            val: 9,
            next: Some(Box::new(ListNode::new(9))),
        }));
        let l2 = Some(Box::new(ListNode::new(1)));
        let result = Solution::add_two_numbers(l1, l2);
        assert_eq!(
            result,
            Some(Box::new(ListNode {
                val: 0,
                next: Some(Box::new(ListNode {
                    val: 0,
                    next: Some(Box::new(ListNode::new(1))),
                })),
            }))
        );
    }
}

File: src/addtwonumbers.nim
type
    ListNode* = ref object
        val*: int
        next*: ListNode

proc printListNode*(l: ListNode) =
    var p = l
    while p != nil:
        echo p.val
        p = p.next

proc newListNode*(val: int): ListNode =
    new(result)
    result.val = val
    result.next = nil

proc addTwoNumbers*(l1: ListNode, l2: ListNode): ListNode =
    var l1 = l1
    var l2 = l2
    var carry = 0
    var dummy = newListNode(0)
    var current = dummy

    while l1 != nil or l2 != nil or carry > 0:
        var sum = carry

        if l1 != nil:
            sum += l1.val
            l1 = l1.next

        if l2 != nil:
            sum += l2.val
            l2 = l2.next
    
        carry = sum div 10
        current.next = newListNode(sum mod 10)
        current = current.next

    return dummy.next
        
proc addListNode*(l: ListNode, val: int) =
    var p = l
    while p.next != nil:
        p = p.next
    p.next = newListNode(val)

proc checkEqual*(l1: ListNode, l2: ListNode): bool =
    var p1 = l1
    var p2 = l2

    while p1 != nil and p2 != nil:
        if p1.val != p2.val:
            return false
        p1 = p1.next
        p2 = p2.next

    return p1 == nil and p2 == nil

# Exporting procedures and types
export addTwoNumbers, ListNode, newListNode, addListNode, checkEqual, printListNode
